File: backend-api/src/services/recommendation-service.ts
import { HealthProfileService } from './health-profile-service'
import { ProductService } from './product-service'
import { ContraindicationRepository } from '../repositories/contraindication-repository'
import { AIServiceClient } from './ai-service-client'
import { NutritionCalculator, CalculationResult } from './nutrition-calculator'
import { Product, ProductWithTranslation } from '../models/product'
import { HealthProfile } from '../models/health-profile'

export interface ProductRecommendation {
  product: ProductWithTranslation
  score: number
  reasons: string[]
  warnings: string[]
  contraindications: Array<{
    id: string
    name: string
    severity: string
  }>
}

export interface RecommendationOptions {
  maxProducts?: number
  excludeProductIds?: string[]
}

export class RecommendationService {
  private healthProfileService: HealthProfileService
  private productService: ProductService
  private contraindicationRepository: ContraindicationRepository
  private aiServiceClient: AIServiceClient
  private nutritionCalculator: NutritionCalculator

  constructor() {
    this.healthProfileService = new HealthProfileService()
    this.productService = new ProductService()
    this.contraindicationRepository = new ContraindicationRepository()
    this.aiServiceClient = new AIServiceClient()
    this.nutritionCalculator = new NutritionCalculator()
  }

  /**
   * Получить рекомендации продуктов для пользователя
   * Использует улучшенный алгоритм оценки с учетом индивидуальных потребностей
   */
  async getRecommendations(
    userId: string,
    options: RecommendationOptions = {}
  ): Promise<ProductRecommendation[]> {
    const profile = await this.healthProfileService.getByUserId(userId)
    if (!profile) {
      throw new Error('Health profile not found')
    }

    // Рассчитываем индивидуальные потребности пользователя
    let nutritionalNeeds: CalculationResult | null = null
    try {
      nutritionalNeeds = this.nutritionCalculator.calculateFullNeeds(profile)
    } catch (error) {
      // Если не удалось рассчитать, продолжаем без индивидуальных потребностей
      console.warn('Could not calculate nutritional needs, using default scoring')
    }

    // Получаем все доступные продукты
    const { products } = await this.productService.getList({ available: true }, 200, 0)

    // Фильтруем исключенные
    const filteredProducts = options.excludeProductIds
      ? products.filter((p) => !options.excludeProductIds!.includes(p.id))
      : products

    // Оцениваем каждый продукт с учетом индивидуальных потребностей
    const recommendations = await Promise.all(
      filteredProducts.map((product) => this.evaluateProduct(product, profile, nutritionalNeeds))
    )

    // Сортируем по оценке и фильтруем противопоказания
    const sorted = recommendations
      .filter((rec) => {
        // Исключаем продукты с критическими противопоказаниями
        const hasCriticalContraindication = rec.contraindications.some(
          (c) => c.severity === 'high'
        )
        return !hasCriticalContraindication
      })
      .sort((a, b) => b.score - a.score)

    // Возвращаем топ продуктов
    const maxProducts = options.maxProducts || 10
    return sorted.slice(0, maxProducts)
  }

  /**
   * Оценить продукт для пользователя
   * Улучшенная версия с учетом индивидуальных потребностей
   */
  private async evaluateProduct(
    product: ProductWithTranslation,
    profile: HealthProfile,
    nutritionalNeeds: CalculationResult | null = null
  ): Promise<ProductRecommendation> {
    let score = 50 // Базовый балл
    const reasons: string[] = []
    const warnings: string[] = []

    // Проверяем противопоказания
    const contraindications = await this.contraindicationRepository.findByProductId(product.id)

    // Проверяем соответствие заболеваний пользователя противопоказаниям
    const userDiseases = profile.diseases || []
    const matchingContraindications = contraindications.filter((contra) => {
      // Упрощенная проверка (в реальности нужна более сложная логика)
      return userDiseases.some((disease) =>
        contra.name_key.toLowerCase().includes(disease.toLowerCase()) ||
        disease.toLowerCase().includes(contra.name_key.toLowerCase())
      )
    })

    if (matchingContraindications.length > 0) {
      // Снижаем оценку за противопоказания
      matchingContraindications.forEach((contra) => {
        if (contra.severity === 'high') {
          score -= 100 // Критическое противопоказание - исключаем продукт
          warnings.push(`High severity contraindication: ${contra.name_key}`)
        } else if (contra.severity === 'medium') {
          score -= 30
          warnings.push(`Medium severity contraindication: ${contra.name_key}`)
        } else {
          score -= 10
          warnings.push(`Low severity contraindication: ${contra.name_key}`)
        }
      })
    }

    // Оценка по цели пользователя (улучшенная с учетом индивидуальных потребностей)
    if (profile.goal) {
      score += this.scoreByGoal(product, profile.goal, reasons, nutritionalNeeds)
    }

    // Оценка по типу продукта и активности
    if (profile.activity_level) {
      score += this.scoreByActivityLevel(product, profile.activity_level, reasons)
    }

    // Оценка по индивидуальным потребностям в макронутриентах
    if (nutritionalNeeds) {
      score += this.scoreByNutritionalNeeds(product, nutritionalNeeds, profile, reasons)
    }

    // Бонус за наличие бренда
    if (product.brand?.verified) {
      score += 5
      reasons.push('Verified brand')
    }

    // Бонус за качество продукта (если есть информация о качестве)
    if (product.brand?.premium) {
      score += 3
      reasons.push('Premium quality product')
    }

    return {
      product,
      score: Math.max(0, score), // Не отрицательный балл
      reasons,
      warnings,
      contraindications: contraindications.map((c) => ({
        id: c.id,
        name: c.name_key,
        severity: c.severity,
      })),
    }
  }

  /**
   * Оценка продукта по цели пользователя
   * Улучшенная версия с учетом индивидуальных потребностей
   */
  private scoreByGoal(
    product: Product,
    goal: string,
    reasons: string[],
    nutritionalNeeds: CalculationResult | null = null
  ): number {
    let score = 0
    const macros = product.macros

    switch (goal) {
      case 'mass':
        // Для набора массы важны белки и калории
        if (product.type === 'protein') {
          score += 20
          reasons.push('High protein for muscle mass gain')
          
          // Дополнительный бонус если белок соответствует индивидуальным потребностям
          if (nutritionalNeeds && macros.protein > 0) {
            const dailyProtein = nutritionalNeeds.protein
            const proteinContribution = macros.protein / dailyProtein
            // Оптимально: 10-25% дневной нормы белка на порцию
            if (proteinContribution >= 0.10 && proteinContribution <= 0.25) {
              score += 8
              reasons.push(`Provides ${Math.round(proteinContribution * 100)}% of daily protein needs`)
            } else if (proteinContribution > 0.25 && proteinContribution <= 0.40) {
              score += 5
            }
          }
        }
        if (product.type === 'creatine') {
          score += 15
          reasons.push('Creatine supports muscle growth')
        }
        // Более точная оценка калорийности для набора массы
        if (macros.calories >= 200 && macros.calories <= 400) {
          score += 10
          reasons.push('Optimal calorie content for mass gain')
        } else if (macros.calories > 400) {
          score += 5
        }
        break

      case 'cut':
        // Для сушки важны белки и низкие калории
        if (product.type === 'protein' && macros.calories < 150) {
          score += 20
          reasons.push('Low-calorie protein for cutting')
        } else if (product.type === 'protein') {
          score += 15
          reasons.push('High protein for cutting')
        }
        if (product.type === 'fat_burner') {
          score += 15
          reasons.push('Fat burner for cutting phase')
        }
        // Штраф за высокую калорийность
        if (macros.calories > 250) {
          score -= 5
        } else if (macros.calories <= 150) {
          score += 8
          reasons.push('Low calorie content')
        }
        if (macros.fats < 5) {
          score += 5
          reasons.push('Low fat content')
        }
        break

      case 'endurance':
        // Для выносливости важны углеводы и аминокислоты
        if (product.type === 'amino') {
          score += 20
          reasons.push('Amino acids for endurance')
        }
        if (product.type === 'pre_workout') {
          score += 15
          reasons.push('Pre-workout for performance')
        }
        // Углеводы важны для выносливости
        if (macros.carbs >= 30 && macros.carbs <= 60) {
          score += 12
          reasons.push('Optimal carb content for energy')
        } else if (macros.carbs > 30) {
          score += 8
          reasons.push('High carb content for energy')
        }
        break

      case 'maintain':
        // Для поддержания сбалансированный состав
        if (product.type === 'protein') {
          score += 15
          reasons.push('Protein for maintenance')
        }
        if (product.type === 'vitamin') {
          score += 10
          reasons.push('Vitamins for overall health')
        }
        // Сбалансированный состав макронутриентов
        const hasProtein = macros.protein > 15
        const hasCarbs = macros.carbs > 10
        if (hasProtein && hasCarbs) {
          score += 5
          reasons.push('Balanced macronutrient profile')
        }
        break
    }

    return score
  }

  /**
   * Оценка продукта по уровню активности
   */
  private scoreByActivityLevel(
    product: Product,
    activityLevel: string,
    reasons: string[]
  ): number {
    let score = 0

    switch (activityLevel) {
      case 'very_high':
      case 'high':
        // Высокая активность - нужны восстановительные продукты
        if (product.type === 'post_workout') {
          score += 15
          reasons.push('Post-workout recovery support')
        }
        if (product.type === 'amino') {
          score += 10
          reasons.push('Amino acids for recovery')
        }
        // Более высокие потребности в белке при высокой активности
        if (product.type === 'protein' && product.macros.protein >= 25) {
          score += 8
          reasons.push('High protein content ideal for high activity')
        }
        break

      case 'moderate':
        // Умеренная активность - стандартные рекомендации
        if (product.type === 'protein') {
          score += 5
        }
        break

      case 'low':
        // Низкая активность - минимум добавок
        if (product.type === 'vitamin') {
          score += 10
          reasons.push('Vitamins for low activity')
        }
        // Низкокалорийные продукты для низкой активности
        if (product.macros.calories < 150) {
          score += 5
        }
        break
    }

    return score
  }

  /**
   * Оценка продукта на основе индивидуальных потребностей в макронутриентах
   */
  private scoreByNutritionalNeeds(
    product: Product,
    nutritionalNeeds: CalculationResult,
    profile: HealthProfile,
    reasons: string[]
  ): number {
    let score = 0
    const macros = product.macros
    const dailyProtein = nutritionalNeeds.protein
    const dailyCalories = nutritionalNeeds.calories

    // Оценка белка: насколько продукт покрывает дневные потребности
    if (macros.protein > 0 && dailyProtein > 0) {
      const proteinContribution = macros.protein / dailyProtein
      
      // Оптимальный диапазон: 10-25% дневной нормы на порцию
      if (proteinContribution >= 0.10 && proteinContribution <= 0.25) {
        score += 12
        reasons.push(`Provides ${Math.round(proteinContribution * 100)}% of daily protein`)
      } else if (proteinContribution > 0.25 && proteinContribution <= 0.40) {
        score += 8
        reasons.push(`High protein: ${Math.round(proteinContribution * 100)}% of daily needs`)
      } else if (proteinContribution >= 0.05 && proteinContribution < 0.10) {
        score += 5
      }
    }

    // Оценка калорийности относительно цели
    if (dailyCalories > 0 && macros.calories > 0) {
      const calorieContribution = macros.calories / dailyCalories
      
      if (profile.goal === 'cut') {
        // Для сушки: низкая калорийность лучше
        if (calorieContribution <= 0.08) {
          score += 8
          reasons.push('Low calorie content suitable for cutting')
        } else if (calorieContribution > 0.15) {
          score -= 5 // Штраф за высокую калорийность при сушке
        }
      } else if (profile.goal === 'mass') {
        // Для набора массы: умеренная-высокая калорийность
        if (calorieContribution >= 0.10 && calorieContribution <= 0.20) {
          score += 8
          reasons.push('Optimal calorie content for mass gain')
        }
      }
    }

    // Оценка баланса макронутриентов
    const totalMacros = macros.protein + macros.carbs + macros.fats
    if (totalMacros > 0) {
      const proteinRatio = macros.protein / totalMacros
      const carbRatio = macros.carbs / totalMacros
      
      // Для белка: оптимальный диапазон 30-50% от макронутриентов
      if (proteinRatio >= 0.30 && proteinRatio <= 0.50) {
        score += 5
      }
      
      // Для углеводов: хороший баланс с белком
      if (profile.goal === 'endurance' && carbRatio >= 0.40) {
        score += 5
        reasons.push('High carb ratio ideal for endurance')
      }
    }

    return score
  }

  /**
   * Проверить совместимость продукта с профилем здоровья
   */
  async checkCompatibility(
    productId: string,
    userId: string
  ): Promise<{ compatible: boolean; warnings: string[]; contraindications: any[] }> {
    const profile = await this.healthProfileService.getByUserId(userId)
    if (!profile) {
      return { compatible: true, warnings: [], contraindications: [] }
    }

    const product = await this.productService.getById(productId)
    if (!product) {
      throw new Error('Product not found')
    }

    const contraindications = await this.contraindicationRepository.findByProductId(productId)
    const userDiseases = profile.diseases || []
    const userMedications = profile.medications || []

    const warnings: string[] = []
    const matchingContraindications = contraindications.filter((contra) => {
      const matchesDisease = userDiseases.some((disease) =>
        contra.name_key.toLowerCase().includes(disease.toLowerCase())
      )
      const matchesMedication = userMedications.some((med) =>
        contra.name_key.toLowerCase().includes(med.toLowerCase())
      )
      return matchesDisease || matchesMedication
    })

    matchingContraindications.forEach((contra) => {
      if (contra.severity === 'high') {
        warnings.push(`CRITICAL: ${contra.name_key} - High severity contraindication`)
      } else {
        warnings.push(`WARNING: ${contra.name_key} - ${contra.severity} severity`)
      }
    })

    const hasCriticalContraindication = matchingContraindications.some(
      (c) => c.severity === 'high'
    )

    return {
      compatible: !hasCriticalContraindication,
      warnings,
      contraindications: matchingContraindications.map((c) => ({
        id: c.id,
        name: c.name_key,
        severity: c.severity,
      })),
    }
  }
}

File: ai-service/app/services/recommendation_service.py
"""
AI-powered recommendation service
Uses ML models to provide personalized product recommendations
"""
from typing import List, Dict, Optional
from datetime import datetime
from app.models.recommendation import (
    ProductRecommendationRequest,
    ProductRecommendationResponse,
    RecommendationsResponse,
)
from app.ml.scoring import ProductScorer
from app.utils.logger import logger
from app.utils.http_client import AsyncHTTPClient


class RecommendationService:
    def __init__(self, backend_api_url: str):
        self.backend_api_url = backend_api_url
        self.client = AsyncHTTPClient(timeout=30.0, max_retries=3, backoff_factor=0.5)
        self.scorer = ProductScorer()

    async def get_ai_recommendations(
        self, request: ProductRecommendationRequest
    ) -> RecommendationsResponse:
        """
        Generate AI-powered product recommendations
        
        Enhanced implementation that:
        1. Fetches base recommendations from backend API
        2. Fetches nutritional needs for personalized scoring
        3. Enhances them with advanced AI scoring
        4. Returns improved recommendations with confidence scores
        """
        try:
            logger.info(f"Generating AI recommendations for user {request.user_id}")
            
            # Fetch base recommendations from backend
            rec_response = await self.client.get(
                f"{self.backend_api_url}/api/v1/recommendations",
                headers={"X-User-ID": request.user_id},
            )
            
            if rec_response.status_code != 200:
                logger.warning(f"Backend API returned {rec_response.status_code}")
                raise Exception(f"Backend API returned {rec_response.status_code}")
            
            data = rec_response.json()
            base_recommendations = data.get("data", [])
            
            if not base_recommendations:
                logger.warning("No base recommendations found")
                return RecommendationsResponse(
                    recommendations=[],
                    generated_at=datetime.utcnow(),
                    user_profile_summary={
                        "goal": request.goal,
                        "activity_level": request.activity_level,
                    },
                )
            
            # Fetch nutritional needs for enhanced scoring
            nutritional_needs = None
            try:
                nutrition_response = await self.client.get(
                    f"{self.backend_api_url}/api/v1/nutrition/calculate",
                    headers={"X-User-ID": request.user_id},
                )
                if nutrition_response.status_code == 200:
                    nutrition_data = nutrition_response.json()
                    if nutrition_data.get("success"):
                        needs_data = nutrition_data.get("data", {})
                        nutritional_needs = {
                            "calories": needs_data.get("calories", 0),
                            "protein": needs_data.get("protein", 0),
                            "carbs": needs_data.get("carbs", 0),
                            "fats": needs_data.get("fats", 0),
                        }
                        logger.info(f"Using nutritional needs: {nutritional_needs}")
            except Exception as e:
                logger.warning(f"Could not fetch nutritional needs: {str(e)}, using defaults")
            
            # Enhance with AI scoring
            enhanced = self._enhance_recommendations(
                base_recommendations, request, nutritional_needs
            )
            
            logger.info(f"Generated {len(enhanced)} enhanced recommendations")
            
            return RecommendationsResponse(
                recommendations=enhanced,
                generated_at=datetime.utcnow(),
                user_profile_summary={
                    "goal": request.goal,
                    "activity_level": request.activity_level,
                    "age": request.age,
                    "gender": request.gender,
                },
            )
            
        except Exception as e:
            logger.error(f"Failed to get recommendations: {str(e)}")
            raise Exception(f"Failed to get recommendations: {str(e)}")

    def _enhance_recommendations(
        self,
        base_recommendations: List[Dict],
        request: ProductRecommendationRequest,
        nutritional_needs: Optional[Dict] = None,
    ) -> List[ProductRecommendationResponse]:
        """
        Enhance recommendations with AI scoring
        
        Args:
            base_recommendations: Base recommendations from backend
            request: Request with user profile data
            nutritional_needs: Optional pre-calculated nutritional needs
        """
        enhanced = []
        
        user_profile = {
            "goal": request.goal,
            "activity_level": request.activity_level,
            "age": request.age,
            "gender": request.gender,
            "weight": request.weight,
            "height": request.height if hasattr(request, "height") else 175,
        }
        
        for rec in base_recommendations:
            product = rec.get("product", {})
            base_score = rec.get("score", 50)
            
            # Calculate enhanced AI score with nutritional needs
            ai_score = self.scorer.calculate_score(
                product,
                user_profile,
                base_score,
                nutritional_needs,
            )
            
            # Calculate confidence with product data
            confidence = self.scorer.calculate_confidence(
                ai_score,
                product.get("type", ""),
                user_profile,
                rec.get("reasons", []),
                product,  # Pass product for brand verification
            )
            
            # Enhance reasons based on AI analysis
            enhanced_reasons = self._generate_ai_reasons(
                product,
                user_profile,
                rec.get("reasons", []),
                nutritional_needs,
            )
            
            enhanced.append(
                ProductRecommendationResponse(
                    product_id=product.get("id", ""),
                    score=ai_score,
                    confidence=round(confidence, 2),
                    reasons=enhanced_reasons,
                    warnings=rec.get("warnings", []),
                    dosage_recommendation=self._suggest_dosage(product, user_profile),
                )
            )
        
        # Sort by AI score (not base score)
        enhanced.sort(key=lambda x: x.score, reverse=True)
        
        # Filter and limit
        max_products = request.max_products or 10
        return enhanced[:max_products]

    def _generate_ai_reasons(
        self,
        product: Dict,
        user_profile: Dict,
        base_reasons: List[str],
        nutritional_needs: Optional[Dict] = None,
    ) -> List[str]:
        """Generate enhanced reasons based on AI analysis with nutritional context"""
        reasons = list(base_reasons)  # Keep base reasons
        
        goal = user_profile.get("goal", "maintain")
        activity_level = user_profile.get("activity_level", "moderate")
        product_type = product.get("type", "")
        weight = user_profile.get("weight", 70)
        macros = product.get("macros", {}) or {}
        
        # Add AI-generated reasons based on goal
        if goal == "mass" and product_type == "protein":
            protein = macros.get("protein", 0)
            if protein >= 25:
                reasons.append(f"High protein content ({protein}g) ideal for muscle growth")
            else:
                reasons.append("High protein content ideal for muscle growth")
        
        if goal == "cut" and product_type == "fat_burner":
            reasons.append("Supports fat loss goals")
        
        if goal == "cut" and macros.get("calories", 0) < 150:
            reasons.append("Low-calorie option perfect for cutting phase")
        
        if activity_level in ["high", "very_high"] and product_type == "creatine":
            reasons.append("Enhances performance for high-intensity training")
        
        if activity_level in ["high", "very_high"] and product_type == "post_workout":
            reasons.append("Accelerates recovery after intense workouts")
        
        # Add reasons based on nutritional needs
        if nutritional_needs:
            daily_protein = nutritional_needs.get("protein", 0)
            product_protein = macros.get("protein", 0)
            if product_protein > 0 and daily_protein > 0:
                contribution = (product_protein / daily_protein) * 100
                if 10 <= contribution <= 25:
                    reasons.append(
                        f"Provides {contribution:.0f}% of daily protein needs per serving"
                    )
        
        # Protein quality reasons
        protein = macros.get("protein", 0)
        if protein >= 25:
            reasons.append("Excellent protein source (25g+)")
        elif protein >= 20:
            reasons.append("High-quality protein source")
        
        # Brand verification
        brand = product.get("brand", {})
        if brand.get("verified", False):
            reasons.append("Verified and trusted brand")
        
        # Remove duplicates and limit
        unique_reasons = []
        seen = set()
        for reason in reasons:
            reason_lower = reason.lower()
            if reason_lower not in seen:
                seen.add(reason_lower)
                unique_reasons.append(reason)
        
        return unique_reasons[:5]  # Limit to 5 most relevant reasons

    def _suggest_dosage(self, product: Dict, user_profile: Dict) -> Dict:
        """Suggest optimal dosage based on product and user profile"""
        product_type = product.get("type", "")
        weight = user_profile.get("weight", 70)
        activity_level = user_profile.get("activity_level", "moderate")
        
        dosage = {
            "frequency": "daily",
            "timing": "post_workout",
            "notes": [],
        }
        
        if product_type == "protein":
            # Protein: 1.6-2.2g per kg bodyweight
            daily_protein_needs = weight * 2.0
            serving_protein = product.get("macros", {}).get("protein", 20)
            servings = max(1, int(daily_protein_needs / (serving_protein * 3)))  # Split across meals
            dosage["servings_per_day"] = servings
            dosage["timing"] = "post_workout and between meals"
        
        elif product_type == "creatine":
            dosage["grams_per_day"] = 5
            dosage["timing"] = "post_workout or before bed"
            dosage["notes"].append("Start with 3-5g daily")
        
        elif product_type == "pre_workout":
            dosage["servings_per_day"] = 1
            dosage["timing"] = "30 minutes before workout"
            dosage["notes"].append("Use only on training days")
        
        elif product_type == "post_workout":
            dosage["servings_per_day"] = 1
            dosage["timing"] = "immediately after workout"
        
        # Adjust for activity level
        if activity_level in ["high", "very_high"]:
            if "servings_per_day" in dosage:
                dosage["servings_per_day"] = int(dosage["servings_per_day"] * 1.2)
        
        return dosage
